### PR TITLE
Refactor sample tags and per-run tags

### PR DESCRIPTION
**Pull request by na--**
This optimizes how sample tags are created, stored and compared in preparation for the cloud output aggregation functionality. There also are some fixes to the per-run tags and system tags and a much more comprehensive test for per-run tags as well.

### CODE REVIEW

File: stats/stats.go [hunk 1]
```diff
@@ -160,11 +161,106 @@ func (t ValueType) String() string {
...
+	for k, v := range data {
+		tags[k] = v
+	}
+	return &SampleTags{tags: tags}
+}
+
+// IntoSampleTags "consumes" the passed map and creates a new SampleTags
+// struct with the data. The map is set to nil as a hint that it shouldn't
+// be changed after it has been transformed into an "immutable" tag set.
+// Oh, how I miss Rust and move semantics... :)
```
Comment (robingustafsson) [inline stats/stats.go]: 😄

File: stats/stats.go [hunk 2]
```diff
@@ -160,11 +161,106 @@ func (t ValueType) String() string {
...
+// UnmarshalJSON deserializes SampleTags from a JSON string.
+func (st *SampleTags) UnmarshalJSON(data []byte) error {
+	if st == nil {
+		*st = SampleTags{}
+	}
+	return json.Unmarshal(data, &st.tags)
+}
+
+// CloneTags copies and sample tags and underlying tag set returns
+// the result. If the receiver is nil, it returns an empty non-nil map.
```
Comment (robingustafsson) [inline stats/stats.go]: Nit pick, but I think the phrasing in this comment can be tweaked to say something like "CloneTags copies the underlying tag set of a sample tags receiver and returns it. If the receiver is nil, it returns an empty non-nil map." perhaps?

File: stats/stats.go [hunk 3]
```diff
@@ -160,11 +161,106 @@ func (t ValueType) String() string {
...
+// present or its value and true if it is.
+func (st *SampleTags) Get(key string) (string, bool) {
+	if st == nil {
+		return "", false
+	}
+	val, ok := st.tags[key]
+	return val, ok
+}
+
+// IsEqual trues to compare two tag sets with maximum efficiency.
```
Comment (robingustafsson) [inline stats/stats.go]: "trues" -> "tries"

File: stats/stats.go (hunk 2 above)
Comment (na--) [reply stats/stats.go]: Totally agree, will fix in a minute

**Pull request merged**